pkg/identity/extensions_test.go

```diff
@@ -0,0 +1,261 @@
+	keys, chain, err := newRevokedLeafChain()
+	chains := [][]*x509.Certificate{chain}
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	err = peertls.AddSignedCertExt(keys[0], chain[0])
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+
+	extMock := new(extensionHandlerMock)
+	verify := func(ext pkix.Extension, chain [][]*x509.Certificate) error {
+		return extMock.verify(ext, chain)
+	}
+
+	handlers := peertls.ExtensionHandlers{
+		{
+			ID:     peertls.ExtensionIDs[peertls.RevocationExtID],
+			Verify: verify,
+		},
+		{
+			ID:     peertls.ExtensionIDs[peertls.SignedCertExtID],
+			Verify: verify,
+		},
+	}
+
+	extMock.On("verify", chains[0][peertls.LeafIndex].ExtraExtensions[0], chains).Return(nil)
+	extMock.On("verify", chains[0][peertls.LeafIndex].ExtraExtensions[1], chains).Return(nil)
+
+	err = handlers.VerifyFunc()(nil, chains)
+	assert.NoError(t, err)
+	extMock.AssertCalled(t, "verify", chains[0][peertls.LeafIndex].ExtraExtensions[0], chains)
+	extMock.AssertCalled(t, "verify", chains[0][peertls.LeafIndex].ExtraExtensions[1], chains)
+	assert.True(t, extMock.AssertExpectations(t))
+
+	// TODO: test error scenario(s)
+}
+
+func TestParseExtensions(t *testing.T) {
+	revokedLeafKeys, revokedLeafChain, err := newRevokedLeafChain()
+	assert.NoError(t, err)
+
+	whitelistSignedKeys, whitelistSignedChain, err := testpeertls.NewCertChain(3)
+	assert.NoError(t, err)
+
+	err = peertls.AddSignedCertExt(whitelistSignedKeys[0], whitelistSignedChain[0])
+	assert.NoError(t, err)
+
+	_, unrelatedChain, err := testpeertls.NewCertChain(1)
+	assert.NoError(t, err)
+
+	tmp, err := ioutil.TempDir("", "TestParseExtensions")
+	if err != nil {
+		t.FailNow()
+	}
+
+	defer func() { _ = os.RemoveAll(tmp) }()
+	revDB, err := NewRevocationDBBolt(filepath.Join(tmp, "revocations.db"))
+	assert.NoError(t, err)
+
+	cases := []struct {
+		testID    string
+		config    peertls.TLSExtConfig
+		extLen    int
+		certChain []*x509.Certificate
+		whitelist []*x509.Certificate
+		errClass  *errs.Class
+		err       error
+	}{
+		{
+			"leaf whitelist signature - success",
+			peertls.TLSExtConfig{WhitelistSignedLeaf: true},
+			1,
+			whitelistSignedChain,
+			[]*x509.Certificate{whitelistSignedChain[2]},
+			nil,
+			nil,
+		},
+		{
+			"leaf whitelist signature - failure (empty whitelist)",
+			peertls.TLSExtConfig{WhitelistSignedLeaf: true},
+			1,
+			whitelistSignedChain,
+			nil,
+			&peertls.ErrVerifyCAWhitelist,
+			nil,
+		},
+		{
+			"leaf whitelist signature - failure",
+			peertls.TLSExtConfig{WhitelistSignedLeaf: true},
+			1,
+			whitelistSignedChain,
+			unrelatedChain,
+			&peertls.ErrVerifyCAWhitelist,
+			nil,
+		},
+		{
+			"certificate revocation - single revocation ",
+			peertls.TLSExtConfig{Revocation: true},
+			1,
+			revokedLeafChain,
+			nil,
+			nil,
+			nil,
+		},
+		{
+			"certificate revocation - serial revocations",
+			peertls.TLSExtConfig{Revocation: true},
+			1,
+			func() []*x509.Certificate {
```

Comment:
Using func closures inside tests usually indicates that it should be a separate test. These make hard to follow what's actually happening in the test and what's being tested and in which order.

Reply:
Also asserts here will end up in the wrong subtest.

Reply:
https://github.com/storj/storj/wiki/Testing#table-tests